scipy/stats/_rvs_sampling.py

```diff
@@ -168,3 +170,258 @@ def rvs_ratio_uniforms(pdf, umax, vmin, vmax, size=1, c=0, random_state=None):
+        rng = qmc_engine or check_random_state(random_state)
+        return self.ppf(rng.random(size))
+
+
+def _fni_input_validation(dist, tol, max_intervals):
+    if int(max_intervals) != max_intervals or max_intervals <= 1:
+        raise ValueError("`max_intervals' must be an integer greater than 1.")
+
+    tol = float(tol)  # if there's an exception, raise it now
+
+    return dist, tol, max_intervals
+
+
+def _fast_numerical_inverse(dist, tol=1e-12, max_intervals=100000):
+    """
+    Generate a fast approximate PPF (inverse CDF) of a probability distribution
+
+    `_fast_numerical_inverse` accepts `dist`, a frozen instance of
+    `scipy.stats.rv_continuous`, and returns a object `fni` with methods
+    that approximate `dist.ppf` and `dist.rvs`. For some distributions,
+    these methods may be faster than those of `dist` itself.
+
+    Parameters
+    ----------
+    dist : scipy.stats.rv_frozen
+        Frozen distribution for which fast approximate PPF is desired
+    tol : float, optional
+        u-error tolerance. The default is 1e-12.
+    max_intervals : int, optional
+        Maximum number of intervals in the cubic Hermite Spline used to
+        approximate the percent point function. The default is 100000.
+
+    Returns
+    -------
+    H: scipy.interpolate.CubicHermiteSpline
+        Interpolant of the distributions's PPF
+    intervals : int
+        The number of intervals of the interpolant
+    midpoint_error : float
+        The maximum u-error at an interpolant interval midpoint
+
+    """
+    dist, tol, max_intervals = _fni_input_validation(dist, tol, max_intervals)
+
+    # [1] Section 2.1: "For distributions with unbounded domain, we have to
+    # chop off its tails at [a] and [b] such that F(a) and 1-F(b) are small
+    # compared to the maximal tolerated approximation error."
+    p = np.array([dist.ppf(tol/10), dist.isf(tol/10)])  # initial interval
+
+    # [1] Section 2.3: "We then halve this interval recursively until
+    # |u[i+1]-u[i]| is smaller than some threshold value, for example, 0.05."
+    while p.size-1 <= np.ceil(max_intervals/2):
+        u = dist.cdf(p)
+        i = np.nonzero(np.diff(u) > 0.05)[0]
+        if not i.size:
+            break
+
+        p_mid = (p[i] + p[i+1])/2
+        p = np.sort(np.concatenate((p, p_mid)))
+
+    # [1] Section 2.3: "Now we continue with checking the error estimate in
+    # each of the intervals and continue with splitting them until [it] is
+    # smaller than a given error bound."
+    while p.size-1 <= max_intervals:
+        # [1] Equation 4-8
+        u = dist.cdf(p)
+        f = dist.pdf(p)
+        H = CubicHermiteSpline(u, p, 1/f)
+
+        # [1] Equation 12
+        u_mid = (u[:-1] + u[1:])/2
+        eu = np.abs(dist.cdf(H(u_mid)) - u_mid)
+
+        i = np.nonzero(eu > tol)[0]
+        if not i.size:
+            break
+
+        p_mid = (p[i] + p[i+1])/2
+        p = np.sort(np.concatenate((p, p_mid)))
```

Comment:
I suppose that without too much trouble I can calculate the pdfs/cdfs at the new grid points (`p_mid`) only. It would be a little slower for distributions with fast `cdfs` because there's some overhead to inserting the results into the `u` and `f` arrays in the right order, but maybe we care more about distributions with the slow `cdfs`.